pkg/utils/utils.go

```diff
@@ -3,164 +3,122 @@ package utils
-				styleSrcDirective = append(styleSrcDirective, sources...)
-			default:
-				klog.Warningf("ignored invalid CSP directive: %v", directive)
-			}
-		}
-	}
-
-	imgSrcDirective = append(imgSrcDirective, data)
-	fontSrcDirective = append(fontSrcDirective, data)
-	scriptSrcDirective = append(scriptSrcDirective, []string{unsafeEval, nonce}...)
-	styleSrcDirective = append(styleSrcDirective, unsafeInline)
-
-	// Construct the full list of directives from the aggregated sources.
-	// This array is a list of directives, where each directive is a string
-	// of the form "<directive-type> <sources>".
-	// The sources are concatenated together with a space separator.
-	// The CSP directives string is returned as a slice of strings, where each string is a directive.
-	return []string{
-		strings.Join(baseUriDirective, " "),
-		strings.Join(defaultSrcDirective, " "),
-		strings.Join(imgSrcDirective, " "),
-		strings.Join(fontSrcDirective, " "),
-		strings.Join(scriptSrcDirective, " "),
-		strings.Join(styleSrcDirective, " "),
-		"frame-src 'none'",
-		"frame-ancestors 'none'",
-		"object-src 'none'",
-	}, nil
-}
-
-func ParseContentSecurityPolicyConfig(csp string) (*map[consolev1.DirectiveType][]string, error) {
-	parsedCSP := &map[consolev1.DirectiveType][]string{}
-	err := json.Unmarshal([]byte(csp), parsedCSP)
-	if err != nil {
-		errMsg := fmt.Sprintf("Error unmarshaling ConsoleConfig contentSecurityPolicy field: %v", err)
-		klog.Error(errMsg)
-		return nil, fmt.Errorf(errMsg)
+	// Append localhost sources if running off-cluster
+	if k8sMode == "off-cluster" {
+		directives[baseURI] = append(directives[baseURI], httpLocalHost, wsLocalHost)
+		directives[defaultSrc] = append(directives[defaultSrc], httpLocalHost, wsLocalHost)
+		directives[imgSrc] = append(directives[imgSrc], httpLocalHost)
+		directives[fontSrc] = append(directives[fontSrc], httpLocalHost)
+		directives[scriptSrc] = append(directives[scriptSrc], httpLocalHost, wsLocalHost)
+		directives[styleSrc] = append(directives[styleSrc], httpLocalHost)
 	}
 
-	// Validate the keys to ensure they are all valid DirectiveTypes
-	for key := range *parsedCSP {
-		// Check if the key is a valid DirectiveType
-		if !isValidDirectiveType(key) {
-			return nil, fmt.Errorf("invalid CSP directive: %v", key)
+	// Merge plugin CSP configuration into directives
+	for directive, sources := range pluginsCSP {
+		switch directive {
+		case defaultSrc:
+			directives[defaultSrc] = append(directives[defaultSrc], sources)
+		case imgSrc:
+			directives[imgSrc] = append(directives[imgSrc], sources)
+		case fontSrc:
+			directives[fontSrc] = append(directives[fontSrc], sources)
+		case scriptSrc:
+			directives[scriptSrc] = append(directives[scriptSrc], sources)
+		case styleSrc:
+			directives[styleSrc] = append(directives[styleSrc], sources)
+		case connectSrc:
+			directives[connectSrc] = append(directives[connectSrc], sources)
+		default:
+			klog.Infof("ignored unsupported CSP directive: %v", directive)
```

Comment:
I think we want to fail on unrecognized values.

Reply:
👍